test/contracts/core/liquidators/twapLiquidator.spec.ts

```diff
@@ -0,0 +1,992 @@
+      }
+
+      it('should clear the auction state', async () => {
+        await subject();
+
+        const auction: any = await liquidator.auctions.callAsync(subjectCaller);
+        expect(auction.orderSize).to.bignumber.equal(ZERO);
+        expect(auction.orderRemaining).to.bignumber.equal(ZERO);
+        expect(auction.lastChunkAuctionEnd).to.bignumber.equal(ZERO);
+        expect(auction.chunkAuctionPeriod).to.bignumber.equal(ZERO);
+        expect(auction.chunkSize).to.bignumber.equal(ZERO);
+        expect(auction.chunkAuction.auction.minimumBid).to.bignumber.equal(ZERO);
+        expect(auction.chunkAuction.auction.startTime).to.bignumber.equal(ZERO);
+        expect(auction.chunkAuction.auction.startingCurrentSets).to.bignumber.equal(ZERO);
+        expect(auction.chunkAuction.auction.remainingCurrentSets).to.bignumber.equal(ZERO);
+        expect(JSON.stringify(auction.chunkAuction.auction.combinedTokenArray)).to.equal(JSON.stringify([]));
+        expect(JSON.stringify(auction.chunkAuction.auction.combinedCurrentSetUnits)).to.equal(JSON.stringify([]));
+        expect(JSON.stringify(auction.chunkAuction.auction.combinedNextSetUnits)).to.equal(JSON.stringify([]));
+        expect(auction.chunkAuction.endTime).to.bignumber.equal(ZERO);
+        expect(auction.chunkAuction.startPrice).to.bignumber.equal(ZERO);
+        expect(auction.chunkAuction.endPrice).to.bignumber.equal(ZERO);
+      });
+
+      describe('when the auction has not been complete', async () => {
+        before(async () => {
+          const minimumBid =  await liquidator.minimumBid.callAsync(liquidatorProxy.address);
+          customBidQuantity = minimumBid;
+        });
+
+        after(async () => {
+          customBidQuantity = undefined;
+        });
+
+        it('should revert', async () => {
+          await expectRevertError(subject());
+        });
+      });
+
+      describe('when the rebalance is not complete', async () => {
+        before(async () => {
+          skipSecondAuction = true;
+        });
+
+        after(async () => {
+          skipSecondAuction = undefined;
+        });
+
+        it('should revert', async () => {
+          await expectRevertError(subject());
+        });
+      });
+    });
+
+    describe('#hasRebalanceFailed', async () => {
+      let subjectSet: Address;
+
+      beforeEach(async () => {
+        subjectSet = liquidatorProxy.address;
+      });
+
+      async function subject(): Promise<boolean> {
+        return liquidator.hasRebalanceFailed.callAsync(subjectSet);
+      }
+
+      it('should return false', async () => {
+        const result = await subject();
+        expect(result).to.equal(false);
+      });
+
+      describe('when the auctionFailPeriod has elapsed', async () => {
+        beforeEach(async () => {
+          await blockchain.increaseTimeAsync(auctionPeriod.mul(2));
+
+          // Enact on-chain timestamp
+          const minimumBid =  await liquidator.minimumBid.callAsync(subjectSet);
```

Comment:
I added `mineBlockAsync` awhile ago to enforce fast forwards

Reply:
love it